constellation-app/constellation
# Created generric JSON config file read/writer.

### Description of the Change

Moved logic that reads/writes JSON configuration files into a generic package in Core Functionality. (genericjsonio).
Use this functionality to support PArameterIOUtlities.java.
Now that a generic JSON config package exists, this can be wrapped up and used to support saving tableview column preferences - ie columns to display, their order, and any column identiifed for sorting.
Updated TableView functionlaity to include a preferences menu item containing a save and load preferences sub-menu. User can now save current table column state for future use - note, only the current table type (ie vertex/transaction) is saved ./ loaded.

Table preferecnes are saved into users constellation directory, within the TableViewPreferences subfolder. The name of the config file saved is prefixed with an indication of the type of data being displayed in the table, however this is hidden from the user.

Additionally updated TableView code to store sort order - which was previously being lost when columns ere manually reordered.

### Alternate Designs

nil

### Why Should This Be In Core?

Provides ability to precan table layouts for specific purposes.

### Benefits

As above

### Possible Drawbacks

I'm not aware of any.

### Verification Process

1. Open a new graph as well as tableview. Add a selection of nodes and transactions to the graph using draw mode (or some other method).
2. Ensure transactions are being displayed in table and update the displayed columns to include non-standard columns. Sort the data by one of the columns
3. Click the heart icon and select 'Save Table Preferences'
4. On the save dialog enter a name - ie 'test' and click Ok.
5. Modify the displayed columns and their order, change the column being sorted by and the direction.
6. Click on the heart icon and select "Load Table Preferences". Select the entry in the list corresponding to the save name entered at step 4 and press the cancel button. **Confirm that no changes are made to the table**.
7. Click on the heart icon and select "Load Table Preferences". Select the entry in the list corresponding to the save name entered at step 4 and press the OK button.  **Confirm that the columns, their order and sorting reflect the changes saved at step 4**
8. Change the table to display vertexes.
9. Make changes to the displayed columns, their order and sorting.
10. Repeat steps 3-7 and confirm functionlaity works for vertexes as it does for transactions. **use the same filename used above, to confirm this is possible**
11. Click on the heart icon and select "Save Table Preferences". In the save dialog hit enter without entering a filename.
12. Navigate to users .CONSTELLATION directory, and then to the TableViewPreferences subdirectory. **Confirm that contained in this folder are filenames transaction-test.json and vertex-test.json (assuming the filename test was used when saving config), confirm an additonal file of the form "<username> at <timestamp>.json exists**

### Applicable Issues

nil
